### ISSUE TITLE
[a11y][web] Selecting menu items triggers a back navigation

### ISSUE THREAD

**Issue by Renzo-Olivares** [flutter/flutter#134842]
Raised internally at b/299847356

**Issue:** Selecting a menu item with semantics enabled, for example through a pop up menu or dropdown menu, triggers a back navigation. This not only dismisses the menu, but also pops the next route from the navigator stack.

**Expected:** Selecting a menu item with semantics enabled dismisses the menu, but does not trigger another route to be popped from the navigator stack.

Bisected to engine roll: https://github.com/flutter/flutter/pull/131897
Bisected to engine commit: https://github.com/flutter/engine/pull/43620

Minimal example for reproduction: https://dartpad.dev/?id=41045409a61245de472cb0083fc94742&channel=master

https://github.com/flutter/flutter/assets/948037/26c9a39f-3ddc-4eb2-bea6-4909d884cc16

<details><summary>Stack trace before change</summary>

```
dart-sdk/lib/_internal/js_dev_runtime/patch/core_patch.dart 941:28                get current
packages/flutter/src/material/popup_menu.dart 347:30                              handleTap
packages/flutter/src/material/ink_well.dart 1188:21                               handleTap
packages/flutter/src/gestures/recognizer.dart 275:24                              invokeCallback
packages/flutter/src/gestures/tap.dart 654:11                                     handleTapUp
packages/flutter/src/gestures/tap.dart 311:5                                      [_checkUp]
packages/flutter/src/gestures/tap.dart 244:7                                      handlePrimaryPointer
packages/flutter/src/gestures/recognizer.dart 630:9                               handleEvent
packages/flutter/src/gestures/pointer_router.dart 98:12                           [_dispatch]
packages/flutter/src/gestures/pointer_router.dart 143:9                           <fn>
dart-sdk/lib/_internal/js_dev_runtime/private/linked_hash_map.dart 21:13          forEach
packages/flutter/src/gestures/pointer_router.dart 141:17                          [_dispatchEventToRoutes]
packages/flutter/src/gestures/pointer_router.dart 127:7                           route
packages/flutter/src/gestures/binding.dart 488:19                                 handleEvent
packages/flutter/src/gestures/binding.dart 468:14                                 dispatchEvent
packages/flutter/src/rendering/binding.dart 439:11                                dispatchEvent
packages/flutter/src/gestures/binding.dart 413:7                                  [_handlePointerEventImmediately]
packages/flutter/src/gestures/binding.dart 376:5                                  handlePointerEvent
packages/flutter/src/gestures/binding.dart 323:7                                  [_flushPointerEventQueue]
packages/flutter/src/gestures/binding.dart 292:9                                  [_handlePointerDataPacket]
lib/_engine/engine/platform_dispatcher.dart 1295:13                               invoke1
lib/_engine/engine/platform_dispatcher.dart 273:5                                 invokeOnPointerDataPacket
lib/_engine/engine/pointer_binding.dart 168:39                                    [_onPointerData]
lib/_engine/engine/pointer_binding.dart 791:20                                    <fn>
lib/_engine/engine/pointer_binding.dart 720:14                                    <fn>
lib/_engine/engine/pointer_binding.dart 317:16                                    loggedHandler
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 574:37  _checkAndCall
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 579:39  dcall
```

</details>

<details><summary>Stack trace after change</summary>

```
dart-sdk/lib/_internal/js_dev_runtime/patch/core_patch.dart 941:28                get current
packages/flutter/src/widgets/navigator.dart 2525:30                               maybePop
packages/flutter/src/widgets/modal_barrier.dart 228:21                            handleDismiss
packages/flutter/src/rendering/proxy_box.dart 4492:24                             [_performTap]
packages/flutter/src/semantics/semantics.dart 3541:14                             <fn>
packages/flutter/src/semantics/semantics.dart 3358:14                             performAction
packages/flutter/src/rendering/binding.dart 447:64                                performSemanticsAction
packages/flutter/src/semantics/binding.dart 119:5                                 [_handleSemanticsActionEvent]
lib/_engine/engine/platform_dispatcher.dart 1295:13                               invoke1
lib/_engine/engine/platform_dispatcher.dart 1173:5                                invokeOnSemanticsAction
lib/_engine/engine/pointer_binding.dart 263:43                                    onClick
lib/_engine/engine/semantics/tappable.dart 34:22                                  <fn>
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 574:37  _checkAndCall
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 579:39  dcall
```
```
dart-sdk/lib/_internal/js_dev_runtime/patch/core_patch.dart 941:28                get current
packages/flutter/src/material/popup_menu.dart 347:30                              handleTap
packages/flutter/src/material/ink_well.dart 1188:21                               handleTap
packages/flutter/src/material/ink_well.dart 861:5                                 simulateTap
packages/flutter/src/rendering/proxy_box.dart 4492:24                             [_performTap]
packages/flutter/src/semantics/semantics.dart 3539:14                             <fn>
packages/flutter/src/semantics/semantics.dart 3357:14                             performAction
packages/flutter/src/rendering/binding.dart 447:64                                performSemanticsAction
```

</details>

From the stack trace it looks like a second semantics action is being dispatched and that is causing a second route to be popped from the navigator stack. The first pop that happens in `PopUpMenu` is expected https://github.com/flutter/flutter/blob/367f9ea16bfae1ca451b9cc27c1366870b187ae2/packages/flutter/lib/src/material/popup_menu.dart#L341-L346 , but the second pop that happens in `ModalBarrier` https://github.com/flutter/flutter/blob/367f9ea16bfae1ca451b9cc27c1366870b187ae2/packages/flutter/lib/src/widgets/modal_barrier.dart#L223-L233 is not expected.

**Comment by yjbanov** [comment]
A few observations:

## Root cause

The root cause is that semantics tree contains nested clickable/tappable nodes:

<img width="1147" alt="Screenshot 2023-09-15 at 4 12 48 PM" src="https://github.com/flutter/flutter/assets/211513/4c7b7558-2641-4979-a945-ffcbe2525573">

With such a DOM tree when a descendant receives a DOM `click` event, its ancestor also receives it. This leads to two `SemanticsAction.tap` events on the framework side, which results in two routes popped instead of one.

## The bug is old

The bug is in the semantics, and it predates this manifestation probably by many stable releases. What happened recently was that what was previously delivered via gesture recognition is now delivered via semantics action, and since the bug is in the semantics it now manifests itself in a new situation.

## The fix

This should be fixable on the web engine side, but clarification is needed from the framework team on what the expected behaviors are in the presence of nested clickable nodes. However, it's also possible that the bug is in the framework. I imagine we don't have nested clickable widgets in the widget tree. My guess is the semantics merging logic ends up merging nodes in this way, and that may not be a good thing to do.

**Comment by HansMuller** [comment]
@chunhtai - please take a look at this

**Comment by chunhtai** [comment]
>  I imagine we don't have nested clickable widgets in the widget tree

We do, but we handle it using gesture arena. For tap gesture the inner most one will win.

> With such a DOM tree when a descendant receives a DOM click event, its ancestor also receives it

Should whichever receives the click that also has semantics tap prevent the event from bubbling?

**Comment by jiahaog** [comment]
Updating the priority to match the internal bug mapping

**Comment by yjbanov** [comment]
I'm reverting the fix for the merged clicks for now.

> Should whichever receives the click that also has semantics tap prevent the event from bubbling?

I'm going to have to test this. One issue I've seen before (which is related to the PR that's being reverted) is that screen readers like to click in the middle of things. If we have overlapping nested clickables, I'm not sure if the screen reader would still end up clicking on the child anyway.

**Comment by chunhtai** [comment]
Yes that is something we should test, if they issue a really click instead of calling the onClick method on the div, this issue will then probably not only limited to flutter.

**Comment by yjbanov** [comment]
Dropping to P1 as the PR that exacerbated the issue has been reverted.
